pysmt/smtlib/annotations.py

```diff
@@ -0,0 +1,94 @@
+#
+# This file is part of pySMT.
+#
+#   Copyright 2014 Andrea Micheli and Marco Gario
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+
+class Annotations(object):
+    """Handles and stores (key,value) annotations for formulae"""
+
+    def __init__(self, initial_annotations=None):
+        if initial_annotations is not None:
+            self._annotations = initial_annotations
+        else:
+            self._annotations = {}
+
+
+    def add(self, formula, annotation, value=None):
+        """Adds an annotation for the given formula, possibly with the
+        specified value"""
+        term_annotations = self._annotations.setdefault(formula, {})
+        values = term_annotations.setdefault(annotation, set())
+        if value is not None:
+            values.add(value)
+
+
+    def remove(self, formula):
```

Comment:
All remove methods assume that formula is defined in annotations. Is this intended?